crates/phactory/src/system/mod.rs

```diff
@@ -834,9 +843,59 @@ impl<Platform: pal::Platform> System<Platform> {
         let contract_key = ContractKey(keypair);
         let contract_pubkey = contract_key.public();
         let contract_info = event.contract_info;
+        let group_id = chain::Hash::from_low_u64_be(contract_info.group_id);
+
         match contract_info.code_index {
-            CodeIndex::NativeCode(_contract_id) => {
-                // TODO(shelven): launch native code instance
+            CodeIndex::NativeCode(contract_id) => {
+                use contracts::*;
+
+                macro_rules! install {
+                    ($contract: expr) => {{
+                        let ecdh_key = contract_key
+                            .derive_ecdh_key()
+                            .expect("Derive ecdh_key should not fail");
+                        install_contract(
+                            &mut self.contracts,
+                            $contract,
+                            contract_key.0.clone(),
+                            ecdh_key,
+                            block,
+                            group_id,
+                        )?
+                    }};
+                }
+
+                let ecdh_key = contract_key
+                    .0
+                    .derive_ecdh_key()
+                    .or(Err(anyhow::anyhow!("Invalid contract key")))?;
+
+                let contract_id = match contract_id {
```

Comment:
I suggest to maintain a counter for native contract deployment. In this way, we can derive the contract hash by `hash(code_id, counter, group_id)` (or something like code_id + caller + caller_nonce, etc). Then we don't need to limit a single deployment in a group. I believe ink deployment is done in this way

Reply:
`ink!` doesn't use counter to deploy multil-instance. It's address is `hash(deployer + code_hash + salt)`.

Now, I implement the native contract by the similar way by adding an extra layer `NativeContractWrapper`.
Now the ids are derived as:

- `pink_contract_id = hash(group_id, deployer, code_hash, salt)`
- `native_contract_id = hash(group_id, deplayer, native_id, salt)`

Reply:
I believer it's addressed by the upstream PR. I feel good to merge this one. Let's take care of the global contract address after the upstream fix.